### PR TITLE
Truncate product description correctly

### LINKED ISSUE #10685: Product description not correctly truncated
## Description

Relates to https://github.com/openfoodfoundation/openfoodnetwork/issues/10537.

If a product description is set without any line breaks, then the three dots `...` are not displayed correctly when truncating it in the shopfront.

**EDIT**

The text is incorrectly truncated not only in width but also in length, as discussed [here](https://openfoodnetwork.slack.com/archives/CDLKH9MM0/p1681333431120579).

## Expected Behavior

A long product description should be truncated correctly (even if it has no line breaks).

## Actual Behaviour

If a product description is set without any line breaks, then the three dots `...` are not displayed correctly when truncating it in the shopfront.

## Steps to Reproduce

**As admin**
0. Log in as an enterprise shop and create shipping/payment methods, and then create two products (1. and 2.) and edit the descriptions respectively:
1. Incorrectly truncated text in **width**: Set `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi venenatis metus diam, eget scelerisque nibh auctor non. Vestibulum tempor non dolor id luctus.` as a product description.
2. Incorrectly truncated text in **length**: Set
`<div>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi venenatis metus diam, eget scelerisque nibh auctor non.</div><p><br/></p><p>Vestibulum tempor non dolor id luctus.<br/></p><p><br/></p>`  as a product description.

3. Create an order cycle and add the the two products to it. If the order cycle is open, these should be visible in the shopfront. More info on the [userguide](https://guide.openfoodnetwork.org/quick-start-guides/producer-shop-quick-setup-guide). 

**As customer**
4. Visit the shopfront: notice the inconsistent text truncation both in width and length

Please note: [this scenario](https://github.com/openfoodfoundation/openfoodnetwork/issues/12692#issuecomment-2242627337) reproduced the S2 regression and requires manual testing of PRs, for this issue.

## Animated Gif/Screenshot

Incorrect truncation in **width**:
image

Incorrect truncation in **length**:
image

## Workaround

Add line-breaks to your product description. Details [here](https://openfoodnetwork.slack.com/archives/CDLKH9MM0/p1681226946425989?thread_ts=1681224973.501529&cid=CDLKH9MM0).

## Severity

bug-s3: a feature is broken but there is a workaround

## Your Environment

* Version used: v4.3.2
* Browser name and version: Firefox 110
* Operating System and version (desktop or mobile): Ubuntu 22.04

### PR DESCRIPTION
**Pull request by sumedhavs**
#### What? Why?

- Closes #10685

If a product description is set without any line breaks, then the three dots ... are not displayed correctly when truncating it in the shopfront.
Both scenarios below are fixed with this solution:

Incorrectly truncated text in width: Set Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi venenatis metus diam, eget scelerisque nibh auctor non. Vestibulum tempor non dolor id luctus. as a product description.

Incorrectly truncated text in length: Set
<div>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi venenatis metus diam, eget scelerisque nibh auctor non.</div><p><br/></p><p>Vestibulum tempor non dolor id luctus.<br/></p><p><br/></p> as a product description.

image

#### What should we test?

Go to home page http://localhost:3000/ and login as an Admin
From the Menu, select Products
Select Edit to edit Carrots product in Fredo's Farm Hub. Set Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi venenatis metus diam, eget scelerisque nibh auctor non. Vestibulum tempor non dolor id luctus. as a product description. Select Update

Select Edit to edit Garlic product in Fredo's Farm Hub. Render 
<div>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi venenatis metus diam, eget scelerisque nibh auctor non.</div><p><br/></p><p>Vestibulum tempor non dolor id luctus.<br/></p><p><br/></p> in HTML. Use the rendered text and update the description. Select Update

The above steps truncate the description prior to the fix in http://localhost:3000/fredo-s-farm-hub/shop:
image

- Visit http://localhost:3000/fredo-s-farm-hub/shop

#### Release notes

Changelog Category: User facing changes | Technical changes

The title of the pull request will be included in the release notes.

#### Dependencies

No dependencies

#### Documentation updates

No documentation updates required

### CODE REVIEW

File: app/webpacker/css/darkswarm/_shop-product-rows.scss [hunk 1]
```diff
@@ -144,20 +144,16 @@
...
           .product-description {
-            white-space: nowrap;
             overflow: hidden;
             text-overflow: ellipsis;
             margin-bottom: 0.75rem;
             cursor: pointer;
+            
             // Force product description to be on one line
-            // and truncate with ellipsis
+            // and truncate with ellipsis
```
Comment (mkllnk) [inline app/webpacker/css/darkswarm/_shop-product-rows.scss:153]: Why did you add this space?
```suggestion
            // and truncate with ellipsis
```

**Review by mkllnk — COMMENTED** [review]
Thank you for this contribution.

Your commit messages seem to include some random text. I'm not sure which editor you use but it's possible to revise the messages.

I'm not sure if the result is correct in all cases. I added a description with a paragraph and your version displays four lines now (three lines of text and one blank).

Before:
Screen Shot 2023-04-19 at 17 11 40

After:
Screen Shot 2023-04-19 at 17 06 47

@filipefurtad0 Is this an improvement or does it need more work?

**Comment by jibees** [comment]
Quick comment: @filipefurtad0 already wrote a pending spec via: #10686

`spec/system/consumer/shopping/products_spec.rb#50`

I think it's great if this PR could remove that pending line and finally solve the spec ;)

File: app/webpacker/css/darkswarm/_shop-product-rows.scss (hunk 1 above)
Comment (sumedhavs) [reply app/webpacker/css/darkswarm/_shop-product-rows.scss:153]: It was unintentional...

Comment (filipefurtad0) [reply app/webpacker/css/darkswarm/_shop-product-rows.scss:153]: > [filipefurtad0](https://github.com/filipefurtad0) Is this an improvement or does it need more work?

Thanks for the pics @mkllnk , they look good :+1: 

To be sure, I've pulled this branch and tested locally with the [strings which reproduce the issue](https://github.com/openfoodfoundation/openfoodnetwork/issues/10685#issue-1663000868):

Width:
image

Length:
image

So, yes, it looks like it fixes the issue :tada:

**Comment by filipefurtad0** [comment]
> Quick comment: @filipefurtad0 already wrote a pending spec via: https://github.com/openfoodfoundation/openfoodnetwork/pull/10686
spec/system/consumer/shopping/products_spec.rb#50
I think it's great if this PR could remove that pending line and finally solve the spec ;)

I did, but it seems the spec fails after pulling this PR (although the PR seems to fix the issue) - this is because Capybara can't find any three dots `...` in the page content, as these seem _not to be_ `html` content:

image

So, it's a bad spec :face_with_head_bandage:

Not sure how we can assert on those three dots... If you have a suggestion @sumedhavs , that would be great, otherwise, happy to try to work out a solution on a separate PR.

**Comment by mkllnk** [comment]
> Not sure how we can assert on those three dots...

I haven't checked but maybe we need to check for the ellipses character, that's a single "letter" containing three dots: `…`

https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow

**Comment by jibees** [comment]
I've rebased PR onto master.

**Comment by jibees** [comment]
I can see this issue is still in development. @sumedhavs Would you still like to work on it?
